torchdata/stateful_dataloader/README.md

```diff
@@ -0,0 +1,215 @@
+# StatefulDataLoader
+
+[**Saving and Loading State**](#saving-and-loading-state) |
+[**Custom State: Map-Style**](#saving-custom-state-with-map-style-datasets) |
+[**Custom State: Iterable-Style**](#saving-custom-state-with-iterable-style-datasets) |
+[**Install guide**](#installation) | [**Beta Usage and Feedback**](#beta-usage-and-feedback) | [**License**](#license)
+
+`StatefulDataLoader` is a drop-in replacement for `torch.utils.data.DataLoader` which offers
+`state_dict/load_state_dict` methods for handling mid-epoch checkpointing which operate on the previous/next iterator
+requested from the dataloader (resp.).
+
+By default, the state includes the number of batches yielded and uses this to naively fast-forward the sampler
+(map-style) or the dataset (iterable-style). However if the sampler and/or dataset include `state_dict/load_state_dict`
+methods, then it will call them during its own `state_dict/load_state_dict` calls. Under the hood, StatefulDataLoader
+handles aggregation and distribution of state across multiprocess workers (but not across ranks).
+
+## Installation
+
+`torchdata.stateful_dataloader` is currently available on the `torchdata` Nightly build.
+
+Using pip:
+
+```bash
+pip install --pre torchdata --extra-index-url https://download.pytorch.org/whl/nightly/cpu
+```
+
+Using conda:
+
+```bash
+conda install torchdata -c pytorch-nightly
+```
+
+## Saving and loading state
+
+```py
+from torchdata.stateful_dataloader import StatefulDataLoader
+...
+
+dataloader = StatefulDataLoader(dataset, num_workers=2, ...)
+for i, batch in enumerate(dataloader):
+  ...
+  if i == 10:
+    state_dict = dataloader.state_dict()
+    break
+...
+
+# Training run resumes with the previous checkpoint
+dataloader = StatefulDataLoader(dataset, num_workers=2, ...)
+# Resume state with DataLoader
+dataloader.load_state_dict(state_dict)
+for i, batch in enumerate(dataloader):
+  ...
+
+```
+
+## Saving Custom State with Map-Style Datasets
+
+For efficient resuming, you can resume iteration by defining `state_dict/load_state_dict` methods in your sampler. If
+your dataset has worker-specific state (eg RNG transform state) you can add `state_dict/load_state_dict` methods to your
+dataset.
+
+```py
+from typing import *
+import torch
+import torch.utils.data
+from torchdata.stateful_dataloader import StatefulDataLoader
+
+class MySampler(torch.utils.data.Sampler[int]):
```

Comment:
Should we mentioned that RandomSampler and BatchSampler has been patched so that it should work out of the box for users who use the default samplers. On a related note, should we also consider making DistributedSampler 'Stateful`?